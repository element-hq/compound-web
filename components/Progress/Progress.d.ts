import { default as React } from '../../../node_modules/react';
type ProgressProps = {
    /** The size variant of the progress bar */
    size: "sm" | "lg";
    /**
     * The colour variant to use for the progress bar indicator
     * If not set, the progress bar will be rendered with a gray tone, which should only be used when the progress bar is empty
     */
    tint?: "red" | "orange" | "lime" | "green";
    /** The CSS class name forwarded to the root element */
    className?: string;
    /** The value of the progress bar. Defaults to max if not provided */
    value?: number | null;
    /** The maximum value of the progress bar. Defaults to 1 if not provided */
    max?: number;
    /**
     * A function to get the text label to display in the progress bar.
     * If set, it will display a label on top of the progress bar
     */
    getValueLabel?: (value: number, max: number) => string;
} & Omit<React.ComponentProps<"div">, "children">;
/**
 * Displays an indicator showing the completion progress of a task, optionally with a label
 */
export declare const Progress: React.ForwardRefExoticComponent<Omit<ProgressProps, "ref"> & React.RefAttributes<HTMLDivElement>>;
export {};
//# sourceMappingURL=Progress.d.ts.map