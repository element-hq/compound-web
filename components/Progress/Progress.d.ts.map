{"version":3,"file":"Progress.d.ts","sourceRoot":"","sources":["../../../src/components/Progress/Progress.tsx"],"names":[],"mappings":"AAOA,OAAO,KAA4B,MAAM,OAAO,CAAC;AAOjD,KAAK,aAAa,GAAG;IACnB,2CAA2C;IAC3C,IAAI,EAAE,IAAI,GAAG,IAAI,CAAC;IAElB;;;OAGG;IACH,IAAI,CAAC,EAAE,KAAK,GAAG,QAAQ,GAAG,MAAM,GAAG,OAAO,CAAC;IAE3C,uDAAuD;IACvD,SAAS,CAAC,EAAE,MAAM,CAAC;IAEnB,qEAAqE;IACrE,KAAK,CAAC,EAAE,MAAM,GAAG,IAAI,CAAC;IAEtB,2EAA2E;IAC3E,GAAG,CAAC,EAAE,MAAM,CAAC;IAEb;;;OAGG;IACH,aAAa,CAAC,EAAE,CAAC,KAAK,EAAE,MAAM,EAAE,GAAG,EAAE,MAAM,KAAK,MAAM,CAAC;CACxD,GAAG,IAAI,CAAC,KAAK,CAAC,cAAc,CAAC,KAAK,CAAC,EAAE,UAAU,CAAC,CAAC;AAElD;;GAEG;AACH,eAAO,MAAM,QAAQ,mGAiDpB,CAAC"}