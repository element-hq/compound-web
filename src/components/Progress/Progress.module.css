/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.progress-bar {
  position: relative;
  background-color: var(--cpd-color-gray-200);
  border: 1px solid var(--cpd-color-gray-400);
  border-radius: var(--cpd-radius-pill-effect);
  overflow: hidden;

  &[data-size="sm"] {
    block-size: var(--cpd-space-2x);
  }

  &[data-size="lg"] {
    block-size: var(--cpd-space-4x);
  }
}

.progress-bar-container {
  display: flex;
  flex-direction: column;
  gap: var(--cpd-space-1x);

  --cpd-progress-bar-main: var(--cpd-color-text-secondary);
  --cpd-progress-bar-muted: var(--cpd-color-gray-800);

  &[data-tint="green"] {
    --cpd-progress-bar-main: var(--cpd-color-text-success-primary);
    --cpd-progress-bar-muted: var(--cpd-color-green-800);
  }

  &[data-tint="lime"] {
    --cpd-progress-bar-main: var(--cpd-color-lime-900);
    --cpd-progress-bar-muted: var(--cpd-color-lime-800);
  }

  &[data-tint="orange"] {
    --cpd-progress-bar-main: var(--cpd-color-orange-900);
    --cpd-progress-bar-muted: var(--cpd-color-orange-800);
  }

  &[data-tint="red"] {
    --cpd-progress-bar-main: var(--cpd-color-text-critical-primary);
    --cpd-progress-bar-muted: var(--cpd-color-red-800);
  }
}

.progress-bar-label {
  font: var(--cpd-font-body-sm-medium);
  letter-spacing: var(--cpd-font-letter-spacing-body-sm);
  color: var(--cpd-progress-bar-main);
}

.progress-bar-indicator {
  position: absolute;
  inset: 0;
  transition: transform 0.2s ease-in-out;
  background-image: linear-gradient(
    135deg,
    var(--cpd-progress-bar-muted) 0%,
    var(--cpd-progress-bar-muted) 25%,
    var(--cpd-progress-bar-main) 25%,
    var(--cpd-progress-bar-main) 50%,
    var(--cpd-progress-bar-muted) 50%,
    var(--cpd-progress-bar-muted) 75%,
    var(--cpd-progress-bar-main) 75%,
    var(--cpd-progress-bar-main) 100%,
    var(--cpd-progress-bar-muted) 100%
  );

  /* sqrt(number of stripes * 2 * (stripe width)^2) = sqrt(4 * 2 * 2^2) = sqrt(32) */
  background-size: 5.6569px 5.6569px;
}
