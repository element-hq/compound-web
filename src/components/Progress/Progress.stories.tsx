/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { Meta, StoryObj } from "@storybook/react-vite";

import { Progress as ProgressComponent } from "./Progress";

function getValueLabel(value: number) {
  if (value === 0) {
    return "Empty progress bar";
  } else if (value <= 25) {
    return "Quarter-full progress bar";
  } else if (value <= 50) {
    return "Half-full progress bar";
  } else if (value <= 75) {
    return "Three-quarters full progress bar";
  } else {
    return "Full progress bar";
  }
}

const meta = {
  title: "Progress",
  component: ProgressComponent,
  tags: ["autodocs"],
  argTypes: {
    max: {
      type: "number",
    },
    value: {
      type: "number",
    },
  },
  args: {
    getValueLabel,
    size: "sm",
    max: 100,
    value: 50,
  },
} satisfies Meta<typeof ProgressComponent>;
export default meta;

type Story = StoryObj<typeof meta>;

export const LargeHalf: Story = {
  args: {
    tint: "green",
    size: "lg",
    value: 50,
  },
};

export const SmallHalf: Story = {
  args: {
    tint: "green",
    size: "sm",
    value: 50,
  },
};

export const LargeFull: Story = {
  args: {
    tint: "green",
    size: "lg",
    value: 100,
  },
};

export const SmallFull: Story = {
  args: {
    tint: "green",
    size: "sm",
    value: 100,
  },
};

export const Empty: Story = {
  args: {
    value: 0,
  },
};

export const Red = {
  args: {
    value: 25,
    tint: "red",
  },
};

export const Orange = {
  args: {
    value: 50,
    tint: "orange",
  },
};

export const Lime = {
  args: {
    value: 75,
    tint: "lime",
  },
};

export const Green = {
  args: {
    value: 100,
    tint: "green",
  },
};
