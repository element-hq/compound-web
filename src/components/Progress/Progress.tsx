/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, { forwardRef, useId } from "react";
import classNames from "classnames";

import styles from "./Progress.module.css";

import { Root, Indicator } from "@radix-ui/react-progress";

type ProgressProps = {
  /** The size variant of the progress bar */
  size: "sm" | "lg";

  /**
   * The colour variant to use for the progress bar indicator
   * If not set, the progress bar will be rendered with a gray tone, which should only be used when the progress bar is empty
   */
  tint?: "red" | "orange" | "lime" | "green";

  /** The CSS class name forwarded to the root element */
  className?: string;

  /** The value of the progress bar. Defaults to max if not provided */
  value?: number | null;

  /** The maximum value of the progress bar. Defaults to 1 if not provided */
  max?: number;

  /**
   * A function to get the text label to display in the progress bar.
   * If set, it will display a label on top of the progress bar
   */
  getValueLabel?: (value: number, max: number) => string;
} & Omit<React.ComponentProps<"div">, "children">;

/**
 * Displays an indicator showing the completion progress of a task, optionally with a label
 */
export const Progress = forwardRef<HTMLDivElement, ProgressProps>(
  function Progress(
    {
      size,
      tint,
      className,
      value: valueProp,
      max: maxProp,
      getValueLabel,
      ...props
    },
    ref,
  ) {
    const max = maxProp ?? 1;
    const value = valueProp ?? max;
    const labelId = useId();
    const label = getValueLabel ? getValueLabel(value, max) : null;

    return (
      <div
        className={classNames(styles["progress-bar-container"], className)}
        data-tint={tint}
        {...props}
      >
        {label && (
          <div id={labelId} className={styles["progress-bar-label"]}>
            {label}
          </div>
        )}

        <Root
          className={styles["progress-bar"]}
          data-size={size}
          max={max}
          value={value}
          ref={ref}
          aria-labelledby={label ? labelId : undefined}
          getValueLabel={getValueLabel}
        >
          <Indicator
            className={styles["progress-bar-indicator"]}
            style={{
              transform: `translateX(-${100 - (value / max) * 100}%)`,
            }}
          />
        </Root>
      </div>
    );
  },
);
