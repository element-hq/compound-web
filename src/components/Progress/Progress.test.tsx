/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, it, expect } from "vitest";
import React from "react";
import { render } from "@testing-library/react";

import * as stories from "./Progress.stories";
import { composeStories } from "@storybook/react";

const { SmallFull, SmallHalf, LargeFull, LargeHalf, Empty, Green, Red, Lime } =
  composeStories(stories);

describe("Progress", () => {
  it("renders a small full progress bar", () => {
    const { container } = render(<SmallFull />);
    expect(container).toMatchSnapshot();
  });

  it("renders a small half progress bar", () => {
    const { container } = render(<SmallHalf />);
    expect(container).toMatchSnapshot();
  });

  it("renders a large full progress bar", () => {
    const { container } = render(<LargeFull />);
    expect(container).toMatchSnapshot();
  });

  it("renders a large half progress bar", () => {
    const { container } = render(<LargeHalf />);
    expect(container).toMatchSnapshot();
  });

  it("renders an empty progress bar", () => {
    const { container } = render(<Empty />);
    expect(container).toMatchSnapshot();
  });

  it("renders a green progress bar", () => {
    const { container } = render(<Green />);
    expect(container).toMatchSnapshot();
  });

  it("renders a red progress bar", () => {
    const { container } = render(<Red />);
    expect(container).toMatchSnapshot();
  });

  it("renders a lime progress bar", () => {
    const { container } = render(<Lime />);
    expect(container).toMatchSnapshot();
  });
});
