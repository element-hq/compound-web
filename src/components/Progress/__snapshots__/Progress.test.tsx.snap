// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Progress > renders a green progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
    data-tint="green"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r5»"
    >
      Full progress bar
    </div>
    <div
      aria-labelledby="«r5»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="100"
      aria-valuetext="Full progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="sm"
      data-state="complete"
      data-value="100"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="complete"
        data-value="100"
        style="transform: translateX(-0%);"
      />
    </div>
  </div>
</div>
`;

exports[`Progress > renders a large full progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
    data-tint="green"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r2»"
    >
      Full progress bar
    </div>
    <div
      aria-labelledby="«r2»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="100"
      aria-valuetext="Full progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="lg"
      data-state="complete"
      data-value="100"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="complete"
        data-value="100"
        style="transform: translateX(-0%);"
      />
    </div>
  </div>
</div>
`;

exports[`Progress > renders a large half progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
    data-tint="green"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r3»"
    >
      Half-full progress bar
    </div>
    <div
      aria-labelledby="«r3»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="50"
      aria-valuetext="Half-full progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="lg"
      data-state="loading"
      data-value="50"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="loading"
        data-value="50"
        style="transform: translateX(-50%);"
      />
    </div>
  </div>
</div>
`;

exports[`Progress > renders a lime progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
    data-tint="lime"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r7»"
    >
      Three-quarters full progress bar
    </div>
    <div
      aria-labelledby="«r7»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="75"
      aria-valuetext="Three-quarters full progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="sm"
      data-state="loading"
      data-value="75"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="loading"
        data-value="75"
        style="transform: translateX(-25%);"
      />
    </div>
  </div>
</div>
`;

exports[`Progress > renders a red progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
    data-tint="red"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r6»"
    >
      Quarter-full progress bar
    </div>
    <div
      aria-labelledby="«r6»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="25"
      aria-valuetext="Quarter-full progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="sm"
      data-state="loading"
      data-value="25"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="loading"
        data-value="25"
        style="transform: translateX(-75%);"
      />
    </div>
  </div>
</div>
`;

exports[`Progress > renders a small full progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
    data-tint="green"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r0»"
    >
      Full progress bar
    </div>
    <div
      aria-labelledby="«r0»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="100"
      aria-valuetext="Full progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="sm"
      data-state="complete"
      data-value="100"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="complete"
        data-value="100"
        style="transform: translateX(-0%);"
      />
    </div>
  </div>
</div>
`;

exports[`Progress > renders a small half progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
    data-tint="green"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r1»"
    >
      Half-full progress bar
    </div>
    <div
      aria-labelledby="«r1»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="50"
      aria-valuetext="Half-full progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="sm"
      data-state="loading"
      data-value="50"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="loading"
        data-value="50"
        style="transform: translateX(-50%);"
      />
    </div>
  </div>
</div>
`;

exports[`Progress > renders an empty progress bar 1`] = `
<div>
  <div
    class="_progress-bar-container_e5fb97"
  >
    <div
      class="_progress-bar-label_e5fb97"
      id="«r4»"
    >
      Empty progress bar
    </div>
    <div
      aria-labelledby="«r4»"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      aria-valuetext="Empty progress bar"
      class="_progress-bar_e5fb97"
      data-max="100"
      data-size="sm"
      data-state="loading"
      data-value="0"
      role="progressbar"
    >
      <div
        class="_progress-bar-indicator_e5fb97"
        data-max="100"
        data-state="loading"
        data-value="0"
        style="transform: translateX(-100%);"
      />
    </div>
  </div>
</div>
`;
